File: impl/js/src/snowflake.ts
export type FromEpoch = number | bigint | Date;
interface SnowflakeGenOpts {
  timestamp?: FromEpoch;
}

interface DeconstructedSnowflake {
  id: bigint;
  timestamp: bigint;
  nodeId: number;
  seq: number;
  epoch: bigint;
}

/**
 * A class for generating and deconstructing snowflakes.
 *
 * Pika has put it's own spin on Twitter snowflakes to simplify deployment
 * and setup. Instead of having a separate worker and process ID, we have
 * one node ID that takes up the 10 bits these fields would usually use.
 *
 * A node ID is computed by taking the MAC address of the first available
 * public interface on the device, then calculating the modulo against
 * 1024 (10b)
 *
 * If we have a snowflake `963584775274749952n` we can represent it as binary:
 * ```
 * 64                                          22           12          0
 *  000011010101111101011111011110000011001010  0001000101  000000000000
 *           number of ms since epoch            node id      sequence
 * ```
 */
export class Snowflake {
  /**
   * Snowflakes generated are derived from this epoch
   * @internal
   */
  #epoch: bigint;

  /**
   * Passed in node ID for this Snowflake instance
   * @internal
   */
  #nodeId: bigint;

  /**
   * Current sequence number (0-4095)
   * @internal
   */
  #seq = 0n;

  /**
   * Last timestamp of the last time the sequence was exhausted
   * @internal
   */
  #lastSequenceExhaustion: number = 0;

  /**
   * @param epoch the base epoch to use
   * @param nodeId optionally pass a static node identifier (0-1023)
   */
  constructor(epoch: FromEpoch, nodeId: number | bigint) {
    this.#epoch = this.normalizeEpoch(epoch);
    this.#nodeId = BigInt(nodeId);
  }

  public get nodeId(): number {
    return Number(this.nodeId);
  }

  public gen({ timestamp = Date.now() }: SnowflakeGenOpts = {}): string {
    const nTimestamp = this.normalizeEpoch(timestamp);

    if (this.#seq === 4095n && timestamp === this.#lastSequenceExhaustion) {
      // purposely blocking
      while (Date.now() - timestamp < 1) {
        continue;
      }
    }

    this.#seq = this.#seq >= 4095n ? 0n : this.#seq + 1n;
    if (this.#seq === 4095n) this.#lastSequenceExhaustion = Date.now();

    return (
      ((nTimestamp - this.#epoch) << 22n) | // millis since epoch
      ((this.#nodeId & 0b1111111111n) << 12n) |
      this.#seq
    ).toString();
  }

  public deconstruct(id: string | bigint): DeconstructedSnowflake {
    const bigIntId = BigInt(id);
    return {
      id: bigIntId,
      timestamp: (bigIntId >> 22n) + this.#epoch,
      nodeId: Number((bigIntId >> 12n) & 0b1111111111n),
      seq: Number(bigIntId & 0b111111111111n),
      epoch: this.#epoch,
    };
  }

  private normalizeEpoch(epoch: FromEpoch): bigint {
    return BigInt(epoch instanceof Date ? epoch.getTime() : epoch);
  }
}

File: impl/py/pika/snowflake.py
from datetime import datetime
from typing import Optional, Union

EpochResolvable = Union[int, datetime]

class DeconstructedSnowflake:
    id: int
    timestamp: int
    node_id: int
    seq: int
    epoch: int

class Snowflake:
    epoch: int
    node_id: int
    seq: int = 0
    last_sequence_exhaustion: int = 0

    def __init__(self, epoch: EpochResolvable, node_id: Union[int, int]) -> None:
        self.epoch = self._normalize_epoch(epoch)
        self.node_id = node_id

    def gen(self, timestamp: Optional[EpochResolvable] = None) -> str:
        n_timestamp = self._normalize_epoch(timestamp or datetime.now())

        if self.seq == 4095 and n_timestamp == self.last_sequence_exhaustion:
            # purposely blocking
            while (datetime.now().timestamp() * 1000) - n_timestamp < 1:
                pass

        self.seq = 0 if self.seq >= 4095 else self.seq + 1
        if self.seq == 4095:
            self.last_sequence_exhaustion = int(datetime.now().timestamp() * 1000)

        return str(((n_timestamp - self.epoch) << 22) | ((self.node_id & 0b1111111111) << 12) | self.seq)

    def deconstruct(self, id: Union[str, int]) -> DeconstructedSnowflake:
        big_int = int(id)

        return DeconstructedSnowflake(
            id=big_int,
            timestamp=(big_int >> 22) + self.epoch,
            node_id=(big_int >> 12) & 0b1111111111,
            seq=big_int & 0b111111111111,
            epoch=self.epoch,
        )

    def _normalize_epoch(self, epoch: EpochResolvable) -> int:
        return int(epoch.timestamp() * 1000) if isinstance(epoch, datetime) else epoch